Media Services version 2015-10-01

BrianBlum opened the pull request:
This is our first draft of a full swagger for Media Services.

Commit:
Media Services version 2015-10-01

azurecla commented:
Hi __@BrianBlum__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (brianbl). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

John-Hart commented:
@BrianBlum can you please start by looking at all of the warnings that the editor.swagger.io is flagging and try to get to a pretty clean state? 

http://editor.swagger.io/#/?Import=https:%2F%2Fraw.githubusercontent.com%2FBrianBlum%2Fazure-rest-api-specs%2F5e6e26edeffb270c3e702bbd1756d46a7559d074%2Farm-mediaservices%2F2015-10-01%2Fswagger%2Fmedia.json

That will really help make the review much faster

BrianBlum commented:
Will do.

From: John Hart [mailto:notifications@github.com]
Sent: Tuesday, June 28, 2016 3:38 PM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Brian Blum <brianbl@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Media Services version 2015-10-01 (#376)


@BrianBlum<https://github.com/BrianBlum> can you please start by looking at all of the warnings that the editor.swagger.io is flagging and try to get to a pretty clean state?

http://editor.swagger.io/#/?Import=https:%2F%2Fraw.githubusercontent.com%2FBrianBlum%2Fazure-rest-api-specs%2F5e6e26edeffb270c3e702bbd1756d46a7559d074%2Farm-mediaservices%2F2015-10-01%2Fswagger%2Fmedia.json<https://na01.safelinks.protection.outlook.com/?url=http%3a%2f%2feditor.swagger.io%2f%23%2f%3fImport%3dhttps%3a%252F%252Fraw.githubusercontent.com%252FBrianBlum%252Fazure-rest-api-specs%252F5e6e26edeffb270c3e702bbd1756d46a7559d074%252Farm-mediaservices%252F2015-10-01%252Fswagger%252Fmedia.json&data=01%7c01%7cbrianbl%40microsoft.com%7c7421bc301da547be37fb08d39fa4d30f%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=CTUYo77T%2bO4U%2bo%2fa4jhPRe64HIZ302WsGUtVO1UIzwc%3d>

That will really help make the review much faster

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-rest-api-specs/pull/376#issuecomment-229205394>, or mute the thread<https://github.com/notifications/unsubscribe/AEH9bMpwvolU6l1PGcZv3JyOHFYwQRVbks5qQaI4gaJpZM4JAhvv>.

BrianBlum commented:
I looked at the link below, and there aren’t any warnings. I also validated the schema using https://json-schema-validator.herokuapp.com/ and validation was successful.

What warnings were you seeing?

From: Brian Blum
Sent: Tuesday, June 28, 2016 4:02 PM
To: 'Azure/azure-rest-api-specs' <reply@reply.github.com>; Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: RE: [Azure/azure-rest-api-specs] Media Services version 2015-10-01 (#376)

Will do.

From: John Hart [mailto:notifications@github.com]
Sent: Tuesday, June 28, 2016 3:38 PM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com<mailto:azure-rest-api-specs@noreply.github.com>>
Cc: Brian Blum <brianbl@microsoft.com<mailto:brianbl@microsoft.com>>; Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Subject: Re: [Azure/azure-rest-api-specs] Media Services version 2015-10-01 (#376)


@BrianBlum<https://github.com/BrianBlum> can you please start by looking at all of the warnings that the editor.swagger.io is flagging and try to get to a pretty clean state?

http://editor.swagger.io/#/?Import=https:%2F%2Fraw.githubusercontent.com%2FBrianBlum%2Fazure-rest-api-specs%2F5e6e26edeffb270c3e702bbd1756d46a7559d074%2Farm-mediaservices%2F2015-10-01%2Fswagger%2Fmedia.json<https://na01.safelinks.protection.outlook.com/?url=http%3a%2f%2feditor.swagger.io%2f%23%2f%3fImport%3dhttps%3a%252F%252Fraw.githubusercontent.com%252FBrianBlum%252Fazure-rest-api-specs%252F5e6e26edeffb270c3e702bbd1756d46a7559d074%252Farm-mediaservices%252F2015-10-01%252Fswagger%252Fmedia.json&data=01%7c01%7cbrianbl%40microsoft.com%7c7421bc301da547be37fb08d39fa4d30f%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=CTUYo77T%2bO4U%2bo%2fa4jhPRe64HIZ302WsGUtVO1UIzwc%3d>

That will really help make the review much faster

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-rest-api-specs/pull/376#issuecomment-229205394>, or mute the thread<https://github.com/notifications/unsubscribe/AEH9bMpwvolU6l1PGcZv3JyOHFYwQRVbks5qQaI4gaJpZM4JAhvv>.

John-Hart commented:
It does take a few minutes for the editor to parse the contents.  The link did work for me but here's one that doesn't have the extra data parameters attached: 

http://editor.swagger.io/#/?Import=https:%2F%2Fraw.githubusercontent.com%2FBrianBlum%2Fazure-rest-api-specs%2F5e6e26edeffb270c3e702bbd1756d46a7559d074%2Farm-mediaservices%2F2015-10-01%2Fswagger%2Fmedia.json.

If you're still not seeing the errors I'll help point them out in the spec for you.

John-Hart commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,679 @@
+      "properties": {
+        "keyType": {
+          "description": "The keyType must be 'Primary' or 'Secondary', indicating which key you want to regenerate.",
+          "type": "string"
```
If Primary and Secondary are the only options should keyType be an enum?

BrianBlum replied:
Yes, this should be an enum. Will update.

From: John Hart [mailto:notifications@github.com]
Sent: Wednesday, June 29, 2016 1:16 PM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Brian Blum <brianbl@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Media Services version 2015-10-01 (#376)


In arm-mediaservices/2015-10-01/swagger/media.json<https://github.com/Azure/azure-rest-api-specs/pull/376#discussion_r69018838>:

> +        },

> +        "storageAccounts": {

> +          "description": "The storage accounts for this resource.",

> +          "type": "array",

> +          "items": {

> +            "$ref": "#/definitions/StorageAccount"

> +          }

> +        }

> +      }

> +    },

> +    "RegenerateKeyInput": {

> +      "description": "The request body for a RegenerateKey API.",

> +      "properties": {

> +        "keyType": {

> +          "description": "The keyType must be 'Primary' or 'Secondary', indicating which key you want to regenerate.",

> +          "type": "string"

If Primary and Secondary are the only options should keyType be an enum?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-rest-api-specs/pull/376/files/5e6e26edeffb270c3e702bbd1756d46a7559d074#r69018838>, or mute the thread<https://github.com/notifications/unsubscribe/AEH9bFyMNOgi4b8veveqDWcKygrGCHusks5qQtKMgaJpZM4JAhvv>.

John-Hart commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,679 @@
+              "$ref": "#/definitions/ApiError"
+            }
+          }
+        }
```
MediaSeriveCollection has a nextLink so is this pageable? If so, consider extension x-ms-pageable.  
See https://github.com/Azure/autorest/blob/master/Documentation/swagger-extensions.md#x-ms-pageable

John-Hart commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,679 @@
+            "schema": {
+              "$ref": "#/definitions/ApiError"
+            }
+          }
```
Is this a Long Running Operation?  If so consider extension x-ms-long-running-operation.
See https://github.com/Azure/autorest/blob/master/Documentation/swagger-extensions.md#x-ms-long-running-operation

John-Hart commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,679 @@
+            "schema": {
+              "$ref": "#/definitions/ApiError"
+            }
+          }
```
Is this a Long Running Operation?  If so consider extension x-ms-long-running-operation.
See https://github.com/Azure/autorest/blob/master/Documentation/swagger-extensions.md#x-ms-long-running-operation

BrianBlum replied:
None of our operations are long running.

From: John Hart [mailto:notifications@github.com]
Sent: Wednesday, June 29, 2016 1:58 PM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Brian Blum <brianbl@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Media Services version 2015-10-01 (#376)


In arm-mediaservices/2015-10-01/swagger/media.json<https://github.com/Azure/azure-rest-api-specs/pull/376#discussion_r69026299>:

> +        "responses": {

> +          "200": {

> +            "description": "OK",

> +            "schema": {

> +              "$ref": "#/definitions/RegenerateKeyOutput"

> +            }

> +          },

> +          "202": {

> +            "description": "Accepted"

> +          },

> +          "default": {

> +            "description": "Detailed error information.",

> +            "schema": {

> +              "$ref": "#/definitions/ApiError"

> +            }

> +          }

Is this a Long Running Operation? If so consider extension x-ms-long-running-operation.
See https://github.com/Azure/autorest/blob/master/Documentation/swagger-extensions.md#x-ms-long-running-operation

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-rest-api-specs/pull/376/files/5e6e26edeffb270c3e702bbd1756d46a7559d074#r69026299>, or mute the thread<https://github.com/notifications/unsubscribe/AEH9bG-qr1QjVnc941qS_HDJH_Fy0L2_ks5qQtxEgaJpZM4JAhvv>.

John-Hart commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,679 @@
+            "schema": {
+              "$ref": "#/definitions/ApiError"
+            }
+          }
```
Is this a Long Running Operation?  If so consider extension x-ms-long-running-operation.
See https://github.com/Azure/autorest/blob/master/Documentation/swagger-extensions.md#x-ms-long-running-operation

John-Hart commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,679 @@
+            "schema": {
+              "$ref": "#/definitions/ApiError"
+            }
+          }
```
Another consideration for Long Running Operation.

John-Hart commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,679 @@
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Media/mediaservices/{mediaServiceName}/listKeys": {
+      "post": {
```
Is this a post or a get?

BrianBlum replied:
Post. Updating swagger.

From: John Hart [mailto:notifications@github.com]
Sent: Wednesday, June 29, 2016 2:08 PM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Brian Blum <brianbl@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Media Services version 2015-10-01 (#376)


In arm-mediaservices/2015-10-01/swagger/media.json<https://github.com/Azure/azure-rest-api-specs/pull/376#discussion_r69027968>:

> +            "maxLength": 24

> +          },

> +          {

> +            "name": "RegenerateKeyInput",

> +            "description": "Properties needed to regenerate the Media Service key.",

> +            "in": "body",

> +            "required": true,

> +            "schema": {

> +              "$ref": "#/definitions/RegenerateKeyInput"

> +            }

> +          }

> +        ]

> +      }

> +    },

> +    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Media/mediaservices/{mediaServiceName}/listKeys": {

> +      "post": {

Is this a post or a get?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-rest-api-specs/pull/376/files/5e6e26edeffb270c3e702bbd1756d46a7559d074#r69027968>, or mute the thread<https://github.com/notifications/unsubscribe/AEH9bD5R6u_6zArQI9y5dJyAW-YJ7gbVks5qQt6bgaJpZM4JAhvv>.

Commit:
Media Services version 2015-10-01

BrianBlum commented:
Addressed issues raised in review.
We have no long running operations.
MediaServicesCollection has been fixed.
KeyType is an enum.
listKeys is a POST.

azurecla commented:
Hi __@BrianBlum__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (brianbl). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

BrianBlum commented:
Need to fix objects to one.

Commit:
Media Services version 2015-10-01

BrianBlum commented:
Fixed objects for CheckNameAvailability and pageable link location.

Commit:
Media Service 2015-10-01

Commit:
Media Service version 2015-10-01

BrianBlum commented:
Fix return code for CheckNameAvailability.
This should be ready for review now.

Commit:
Media Service version 2015-10-01

BrianBlum commented:
Make description for 200 return codes consistent

Commit:
Media Services version 2015-10-01 with review feedback

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Media/mediaservices/{mediaServiceName}": {
+      "get": {
+        "summary": "Get a Media Service.",
+        "operationId": "MediaServices_GetMediaService",
```
"MediaServices_GetMediaService". There is no point in repeating the noun for the method name. For the operationId we like to follow the Noun_Verb pattern as much as possible. This can be effectively named as "MediaServices_Get".
The generated code looks better **client.MediaServices.Get()**

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+      },
+      "put": {
+        "summary": "Create a Media Service.",
+        "operationId": "MediaServices_PutMediaService",
```
"MediaServices_Create"

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+          "201": {
+            "description": "Created",
+            "schema": {
+              "$ref": "#/definitions/MediaService"
```
Does the Service follow a workflow where the client is supposed to poll till the provisioningState received in 201 reaches a terminal state? If so then this is a long running operation. Please add the x-ms-long-running-operation extension over here.

amarzavery replied:
If it is not a long running operation then ignore this. I just read your comment that "none of the operations are long running".

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+      },
+      "delete": {
+        "summary": "Delete a Media Service.",
+        "operationId": "MediaServices_DeleteMediaService",
```
"MediaServices_Delete"

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+      },
+      "patch": {
+        "summary": "Update a Media Service.",
+        "operationId": "MediaServices_UpdateMediaService",
```
"MediaServices_Update"

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+      "description": "The request body for CheckNameAvailability API.",
+      "properties": {
+        "name": {
+          "description": "Specifies the type of the resource.",
```
"Specifies the name of the mediaservice to check whether it is available". It would be nice to add the same validation constraints (regex, min and max length) that are provided for the mediaServiceName path parameter. It is better to fail fast rather than saying that the desired name is available and then fail on validation when the user is creating a media service.

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+      "description": "The request body for CheckNameAvailability API.",
+      "properties": {
+        "nameAvailable": {
+          "description": "Specifiies if the name is available.",
```
Specif **~~i~~** ies

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+          "type": "boolean"
+        },
+        "reason": {
+          "description": "Specifiies the reason if the name is not available",
```
Specif **~~i~~** ies

amarzavery commented on arm-mediaservices/2015-10-01/swagger/media.json:
```diff
@@ -0,0 +1,715 @@
+    "MediaService": {
+      "description": "The properties of a Media Service resource.",
+      "properties": {
+        "properties": {
```
It is recommended to use "x-ms-client-flatten" extension over here. This will bring the properties of MediaServiceProperties one level up during deserialization and down while serializing it, thus making it easier for customers.
Take a look at this [Storage example](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-storage/2016-01-01/swagger/storage.json#L760).
More info about this extension can be found [here](https://github.com/Azure/autorest/blob/master/Documentation/swagger-extensions.md#x-ms-client-flatten).

amarzavery commented:
@BrianBlum - Any updates on the comments in the review?

BrianBlum commented:
Thanks for the feedback. I’ve been away on another deployment issue and will return to this now.

From: Amar Zavery [mailto:notifications@github.com]
Sent: Wednesday, July 6, 2016 2:27 PM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Brian Blum <brianbl@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Media Services version 2015-10-01 (#376)


@BrianBlum<https://github.com/BrianBlum> - Any updates on the comments in the review?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-rest-api-specs/pull/376#issuecomment-230912153>, or mute the thread<https://github.com/notifications/unsubscribe/AEH9bMIsJvWiFme1J0RzVAWKqzm1N18Nks5qTB2YgaJpZM4JAhvv>.

Commit:
Media Services version 2015-10-01. Review feedback

BrianBlum commented:
Incorporated review feedback - OperationIds, spelling and flatten the MediaService object.

amarzavery commented:
LGTM. Shall merge once the CI passes.

BrianBlum commented:
Thanks Amar.

AutorestCI commented:
No modification for Python

AutorestCI commented:
No modification for NodeJS
